examples/dataflow-hashpipeline/hashpipeline.py

```diff
@@ -0,0 +1,213 @@
+from apache_beam.options.pipeline_options import PipelineOptions
+
+# NOTE: In the latest versions of Beam for Python there are several issues
+# wth saving the main session that cause bugs in dataflow, so instead of
+# requiring at the top of the file, we do so inside the DoFns and PTransforms
+# See: BEAM-10150 for more
+
+class HashPipelineOptions(PipelineOptions):
+  @classmethod
+  def _add_argparse_args(cls, parser):
+    parser.add_argument('--input_subscription',
+      help='Fully qualified input Pubsub subscription that forwards raw events from GCS. i.e. projects/*/subscriptions/*')
+    parser.add_argument('--output_topic',
+      help='Fully qualified output Pubsub topic name. i.e. projects/*/topics/*')
+    parser.add_argument('--secret_name', type=cls.valid_secret,
+      help='Fully qualified name of the Secret Maneger secret where the base64 encoded hash key is stored. i.e. projects/*/secrets/*"')
+    parser.add_argument('--firestore_project',
+      help='Project Id where Firestore database is located and where DLP will be executed.')
+    parser.add_argument('--collection_name',
+      help='Name of the Firestore collection where hashed SSNs are stored')
+    parser.add_argument('--salt',
+      help='Random string for salting hashes.')
+
+  @classmethod
+  def valid_secret(cls, secret):
+    parts = secret.split('/')
+    if len(parts) != 4 or parts[0] != 'projects' or parts[2] != 'secrets':
+      raise ArgumentTypeError("Secret Name must be of format: projects/<PROJECT>/secrets/<SECRET_NAME>")
+    return secret
+
+def run(argv):
+  """Main entry point; defines and runs the pipeline."""
+  opts = HashPipelineOptions()
+  std_opts = opts.view_as(StandardOptions)
+  std_opts.streaming = True 
+
+  def fmt(elem):
+    filename, counts = elem
+    obj = {'filename': filename, 'ssn_counts': counts}
+    return json.dumps(obj).encode('utf-8')
+
+  with beam.Pipeline(options=opts) as p:
+    (p | 'Read filename from Pubsub' >> ReadFromPubSub(None, opts.input_subscription)
+       | 'Decode from UTF-8' >> beam.Map(lambda elem: elem.decode('utf-8'))
+       | 'Read filename and lines from file' >> ReadAllFromTextWithFilename()
+       | 'Get DLP Findings' >> beam.ParDo(DlpFindingDoFn(opts.firestore_project, std_opts.runner))
+       | 'Determine if client SSN' >> beam.ParDo(ExistingSSNsDoFn(
+          opts.firestore_project, opts.salt, opts.secret_name, opts.collection_name, std_opts.runner))
+       | 'Apply windowing' >> beam.WindowInto(FixedWindows(5))
+       | 'Group by Filename' >> beam.GroupByKey()
+       | 'Count SSNs by Filename' >> beam.Map(lambda elem : (elem[0], sum(elem[1])))
+       | 'Format Message' >> beam.Map(fmt)
+       | 'Write to Pubsub' >> WriteToPubSub(opts.output_topic))
+
+class DlpFindingDoFn(beam.DoFn):
+  """Fetch DLP Findings as a PCollection"""
+  def __init__(self, project, runner):
+    beam.DoFn.__init__(self)
+    self.project = project
+    self.runner = runner
+    self.inspect_config = {
```

Comment:
Since this example is specific to Social Security Numbers, what benefit would there be in referencing a template?

Reply:
I 100% agree w/ theo that the template should be a parameter.
It would make this a much more generally applicable example and thus have broader impact.

When possible it's best to make these pipelines thin layers that wrap the APIs they're calling for enhancement exposing as many parameters as possible from the underlying API to the user (within reason) and choosing reasonable defaults based on your current use case. 

This design approach avoids over fitting to the current use case and expose a surface for re-use for un-foreseen use cases without code change.

For very little effort, you could make this one stop shopping for "compare to my known X hashed dictionary".

I could imagine this pattern would be identical for other InfoTypes namely:

1. SWIFT ID (your current / motivating customer for this asset may even be interested in this :) )
2. Credit Card numbers
3. Generic IDs (e.g. a user id)
4. MAC Address
5. Person Name
6. Phone Number

Reply:
Its really for the flexibility. Alternatively if you want make the inspect config a global variable instead of hardcoding it in the init method, that would could accomplish the same thing.